### PR TITLE
Add AdoptedResource functionality to firewall

### PR DESCRIPTION
**Pull request by nnbu**
Issue #: https://github.com/aws-controllers-k8s/community/issues/1947

RCA:

Following CR currently fails to adopt the existing firewall resource.
```
apiVersion: services.k8s.aws/v1alpha1
kind: AdoptedResource
metadata:
  name: fw1
spec:
  aws:
    nameOrID: myfw1
  kubernetes:
    group: networkfirewall.services.k8s.aws
    kind: Firewall
    metadata:
      name: fw-adopted
      namespace: default
```

The error comes from https://github.com/aws-controllers-k8s/runtime/blob/main/pkg/runtime/adoption_reconciler.go#L157

`SetIdentifiers` function needs to have code to fill up `required` values of the CR so as to solve this issue.

Solution:
4 new fileds are defined as `firewallPolicyARN`, `vpcID`, `numSubnetMappings`, `subnetMapping{num}` where `num` starts from 0. These fields need to part of `additionalKeys` field of `AdoptedResource` CR for `Firewall`.

The new CR looks like,

```
apiVersion: services.k8s.aws/v1alpha1
kind: AdoptedResource
metadata:
  name: fw1
spec:
  aws:
    nameOrID: myfw1
    additionalKeys:
      firewallPolicyARN: fw-policy-arn
      vpcID: vpc-xxxxx
      numSubnetMappings: "1"
      subnetMapping0: subnet-xxxxx
  kubernetes:
    group: networkfirewall.services.k8s.aws
    kind: Firewall
    metadata:
      name: sg-adopted
      namespace: default
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by nnbu** [comment]
@a-hilaly Could you please review when you get a chance?

File: pkg/resource/firewall/hook.go [hunk 1]
```diff
@@ -27,3 +31,32 @@ var (
...
+		if num > 0 {
+			for i := range numSubnetMappings {
+				key := "subnetMapping" + fmt.Sprint(i)
+				if subnetMapping, ok := identifier.AdditionalKeys[key]; ok {
+					sm := svcapitypes.SubnetMapping{
+						SubnetID: &subnetMapping,
+					}
+					r.ko.Spec.SubnetMappings = append(r.ko.Spec.SubnetMappings, &sm)
+				}
+			}
```
Comment (a-hilaly) [inline pkg/resource/firewall/hook.go:54]: Does this mean that the user should privde a `subnetMapping0` or `subnetMapping1`? Not a crazy user experience :/

File: pkg/resource/firewall/hook.go
Comment (a-hilaly) [inline pkg/resource/firewall/hook.go:1]: Can we have a some e2e tests to validate that thiis works?

File: pkg/resource/firewall/hook.go (hunk 1 above)
Comment (nnbu) [reply pkg/resource/firewall/hook.go:54]: User needs to provide `numSubnetMappings` to indicate how many subnet mappings are there. and then provide the actual subnet mappings starting from `subnetMapping0`.

Please let me know if we can do something better.

Comment (a-hilaly) [reply pkg/resource/firewall/hook.go:54]: I think the best is to introduce `runtime.RawExtension` to a new field called `AdditionnalRawKeys` where users can provide any type and we'll have the responsibility parse it properly

File: pkg/resource/firewall/hook.go
Comment (nnbu) [reply pkg/resource/firewall/hook.go:1]: I tried to find an example of `adoptedResources` e2e tests in any of the repos. But I don't think, there is any. We will need to create a firewall resource upfront and then adopt it using `AdoptedResource` CR.

Do you have a suggestion?

**Comment by nnbu** [comment]
I accidentally updated my branch with upstream and ended up deleting the changes, which closed this PR

**Pull request closed without merging**